### PR TITLE
Add design principles vignette

### LINKED ISSUE #34: Add package design vignette
This has already been authored locally and just needs to be uploaded as a vignette.

### PR DESCRIPTION
**Pull request by jamesmbaazam**
This PR closes #34 by adding a design vignette to the package website.

### CODE REVIEW

**Comment by jamesmbaazam** [comment]
@sbfnk this was sitting on my computer and is definitely out of date, but you could still give it a quick glance. I intend to give it a substantive revision.

File: vignettes/design-principles.Rmd [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+The goal here is to make it easy to acquaint oneself with the code base
+in the absence of the current maintainer. This will ease future contributions
+and maintenance of the package.
+
+## Scope
+
+{epichains} aims to facilitate:
+
+* Simulation of transmission chains using branching processes methods,
+* Handling of transmission chains using objected-orientated design, and
```
Comment (sbfnk) [inline vignettes/design-principles.Rmd:34]:
```suggestion
* Handling of transmission chains with an easily accessible user interface, and
```

I think OOP is more lower level

File: vignettes/design-principles.Rmd [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+The `<epichains>` class extends the `<data.frame>`, using columns
+to store information about the simulated transmission chains and the
+parameter values as attributes. `<data.frame>` was chosen because its tabular
+structure allows us to store information in rows and columns, and is a widely
+used data structure in R. Similarly, the `<epichains_summary>` class is a
+superclass (an extension) of R's `<numeric>` class and stores the parameter
+values as attributes.
+
+The `<epichains>` object contains information about the whole outbreak, but
```
Comment (sbfnk) [inline vignettes/design-principles.Rmd:73]:
```suggestion
The `<epichains>` object contains information about the whole simulated outbreak (potentially with multiple simulations), but
```

File: vignettes/design-principles.Rmd [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+key summaries are not easily deduced from a quick glance of the object. Hence,
+the class has a dedicated `format()/print()` method to print the simulated
+transmission chains in a manner similar to a `<dataframe>`, but accompanied
+by extra summary information including the number of chains simulated, number
+of generations reached, and the number of unique infectors created. These
+summaries are useful for quickly assessing the output of the simulation.
+
+Importantly, the `<epichains>` class has a `summary()` method that returns
```
Comment (sbfnk) [inline vignettes/design-principles.Rmd:81]:
```suggestion
The `<epichains>` class has a `summary()` method that returns
```

File: vignettes/design-principles.Rmd [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+Lastly, `<epichains>` objects have an `aggregate()` method that aggregates the
+transmission tree into cases by "generation" or "time". This is syntactic sugar
+for the `dplyr::group_by()` and `dplyr::summarise()` style of aggregation.
+
+In summary, an `<epichains>` object has the following structure:
+
+* Columns:
+  * `infectee_id` (`<numeric>`)
+  * `infector_id` (`<numeric>`)
+  * `generation` (`<numeric>`), and optionally,
```
Comment (sbfnk) [inline vignettes/design-principles.Rmd:97]: are these numeric or integer? (really they should be integer)

**Review by sbfnk — CHANGES_REQUESTED** [review]
This is great! Only some small suggestions.

As a general point for text-based files such as this one or `.tex` files I'm a big fan of the "one line per sentence" (rather than 80 character cutoff) format as it makes changes easier to apply/visualise in a diff-based interface such as the one used by github. But everyone has different preferences/editors etc. so I don't think there's a hard and fast rule.

File: vignettes/design-principles.Rmd (hunk 4 above)
Comment (jamesmbaazam) [reply vignettes/design-principles.Rmd:97]: They should be. They start off as integer but thanks to R's type instability, they end up being numeric. I think there's an issue to resolve this here: https://github.com/epiverse-trace/epichains/issues/136.

Comment (jamesmbaazam) [inline vignettes/design-principles.Rmd:97]:
```suggestion
  * `infectee_id` (`<integer>`)
  * `infector_id` (`<integer>`)
  * `generation` (`<integer>`), and optionally,
```

**Comment by jamesmbaazam** [comment]
> I'm a big fan of the "one line per sentence" (rather than 80 character cutoff) format as it makes changes easier to apply/visualise in a diff-based interface such as the one used by github. But everyone has different preferences/editors etc. so I don't think there's a hard and fast rule.

Thanks. It's a good point. I'll do that in the future and make sure R Studio wraps it (for me) to make it easier to work with.

**Comment by jamesmbaazam** [comment]
Merging this as it has already been reviewed.

**Pull request merged**